machine/v1/types_nutanixprovider.go

```diff
@@ -0,0 +1,117 @@
+	// +kubebuilder:validation:Minimum=1
+	NumSockets int32 `json:"numSockets"`
+
+	// memorySize is the memory size (in Quantity format) of the VM
+	// The minimum memorySize is 2Gi bytes
+	// +kubebuilder:validation:Required
+	MemorySize resource.Quantity `json:"memorySize"`
+
+	// systemDiskSize is size (in Quantity format) of the system disk of the VM
+	// The minimum systemDiskSize is 20Gi bytes
+	// +kubebuilder:validation:Required
+	SystemDiskSize resource.Quantity `json:"systemDiskSize"`
+
+	// userDataSecret is a local reference to a secret that contains the
+	// UserData to apply to the VM
+	UserDataSecret *corev1.LocalObjectReference `json:"userDataSecret,omitempty"`
+
+	// credentialsSecret is a local reference to a secret that contains the
+	// credentials data to access Nutanix PC client
+	// +kubebuilder:validation:Required
+	CredentialsSecret *corev1.LocalObjectReference `json:"credentialsSecret"`
+}
+
+// NutanixIdentifierType is an enumeration of different resource identifier types.
+type NutanixIdentifierType string
+
+const (
+	// NutanixIdentifierUUID is a resource identifier identifying the object by UUID.
+	NutanixIdentifierUUID NutanixIdentifierType = "uuid"
+
+	// NutanixIdentifierName is a resource identifier identifying the object by Name.
+	NutanixIdentifierName NutanixIdentifierType = "name"
+)
+
+// NutanixResourceIdentifier holds the identity of a Nutanix PC resource (cluster, image, subnet, etc.)
+// +union
+type NutanixResourceIdentifier struct {
+	// Type is the identifier type to use for this resource.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=uuid;name
+	Type NutanixIdentifierType `json:"type"`
+
+	// uuid is the UUID of the resource in the PC.
+	// +optional
+	UUID *string `json:"uuid,omitempty"`
+
+	// name is the resource name in the PC
+	// +optional
+	Name *string `json:"name,omitempty"`
+}
+
+// NutanixMachineProviderStatus is the type that will be embedded in a Machine.Status.ProviderStatus field.
+// It contains nutanix-specific status information.
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type NutanixMachineProviderStatus struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// conditions is a set of conditions associated with the Machine to indicate
+	// errors or other status
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// vmUUID is the Machine associated VM's UUID
+	// The field is missing either before the VM is created, or after the VM is deleted
+	// Once the VM is created, the field is filled with the VM's UUID and it will not change
+	// until the VM is deleted. The vmUUID is used to find the VM when updating the Machine status,
+	// and to delete the VM when the Machine is deleted.
+	// +optional
+	VmUUID *string `json:"vmUUID,omitempty"`
```

Comment:
Given the Machine maps 1:1 with a VM in Nutanix, I don't think we want to clear the UUID when the VM is deleted. What's the motivation behind removing the UUID from the status here? In other providers that do this, we add it, and then never remove/change it

Reply:
There is a scenario that the Machine's associated VM is deleted not via deleting the Machine CR (but deleted from the PC Console or by calling the PC APIs). In this case, when the mapi-nutanix controller updating the Machine CR's status, it need to clear this vmUUID field, besides setting the corresponding status condition. In this scenario, the best solution would be for the Cloud Controller Manager (CCM) node controller to detect the node and VM is gone and then delete the corresponding Machine CR. Since Nutanix does not provide a CCM node controller yet, at least we need to make the Machine CR's status reflect the VM is deleted by clearing the vmUUID field and setting the corresponding status condition.

Reply:
In this scenario, I would expect the `Exists` call within the actuator to fail. If the machine is already provisioned or running, this moves the Machine into a failed phase. At which point, we also put a condition on the Machine to say that the instance no longer exists. 

I don't think we need this additional logic directly within the nutanix actuator, I would expect this field to stay populated.

Perhaps you can expand on what you're trying to achieve by removing the field value? Is something else relying on that?